api/operator/v1alpha1/vmdistributedcluster_types.go

```diff
@@ -0,0 +1,458 @@
+	// +optional
+	TenantID *string `json:"tenantID,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+// CustomVMAgentSpec is a customized specification of a new VMAgent.
+// It includes selected options from the original VMAgentSpec.
+type CustomVMAgentSpec struct {
+	// ParsingError contents error with context if operator was failed to parse json object from kubernetes api server
+	ParsingError string `json:"-" yaml:"-"`
+	// PodMetadata configures Labels and Annotations which are propagated to the vmagent pods.
+	// +optional
+	PodMetadata *vmv1beta1.EmbeddedObjectMetadata `json:"podMetadata,omitempty"`
+	// ManagedMetadata defines metadata that will be added to the all objects
+	// created by operator for the given CustomResource
+	ManagedMetadata *vmv1beta1.ManagedObjectsMetadata `json:"managedMetadata,omitempty"`
+	// LogLevel for VMAgent to be configured with.
+	// INFO, WARN, ERROR, FATAL, PANIC
+	// +optional
+	// +kubebuilder:validation:Enum=INFO;WARN;ERROR;FATAL;PANIC
+	LogLevel string `json:"logLevel,omitempty"`
+	// LogFormat for VMAgent to be configured with.
+	// +optional
+	// +kubebuilder:validation:Enum=default;json
+	LogFormat string `json:"logFormat,omitempty"`
+
+	// RemoteWrite list of victoria metrics /some other remote write system
+	// for vm it must looks like: http://victoria-metrics-single:8428/api/v1/write
+	// or for cluster different url
+	// https://docs.victoriametrics.com/victoriametrics/vmagent/#splitting-data-streams-among-multiple-systems
+	// +optional
+	RemoteWrite []CustomVMAgentRemoteWriteSpec `json:"remoteWrite"`
+	// RemoteWriteSettings defines global settings for all remoteWrite urls.
+	// +optional
+	RemoteWriteSettings *vmv1beta1.VMAgentRemoteWriteSettings `json:"remoteWriteSettings,omitempty"`
+
+	// ShardCount - numbers of shards of VMAgent
+	// in this case operator will use 1 deployment/sts per shard with
+	// replicas count according to spec.replicas,
+	// see [here](https://docs.victoriametrics.com/victoriametrics/vmagent/#scraping-big-number-of-targets)
+	// +optional
+	ShardCount *int `json:"shardCount,omitempty"`
+
+	// UpdateStrategy - overrides default update strategy.
+	// works only for deployments, statefulset always use OnDelete.
+	// +kubebuilder:validation:Enum=Recreate;RollingUpdate
+	// +optional
+	UpdateStrategy *appsv1.DeploymentStrategyType `json:"updateStrategy,omitempty"`
+	// RollingUpdate - overrides deployment update params.
+	// +optional
+	RollingUpdate *appsv1.RollingUpdateDeployment `json:"rollingUpdate,omitempty"`
+	// PodDisruptionBudget created by operator
+	// +optional
+	PodDisruptionBudget       *vmv1beta1.EmbeddedPodDisruptionBudgetSpec `json:"podDisruptionBudget,omitempty"`
+	*vmv1beta1.EmbeddedProbes `json:",inline"`
+	// DaemonSetMode enables DaemonSet deployment mode instead of Deployment.
+	// Supports only VMPodScrape
+	// (available from v0.55.0).
+	// Cannot be used with statefulMode
+	// +optional
+	DaemonSetMode bool `json:"daemonSetMode,omitempty"`
```

Comment:
As far as I understand, it's only possible to run vmagent for replication only at stateful mode. Otherwise, it will lose buffered data on restarts.

 It also would be great to have storageSpec for it required as well as maxOnDiskBuffer size ( aka maxDiskUsagePerURL)